utils/_context/_scenarios/__init__.py

```diff
@@ -165,16 +174,24 @@ def _get_warmups(self):
         if not self.replay:
             warmups.append(create_network)
 
-            for container in self._required_containers:
-                warmups.append(container.start)
-
-        for container in self._required_containers:
-            warmups.append(container.post_start)
+        warmups.append(self._start_containers)
 
         return warmups
 
+    def _start_containers(self):
+        required_thread = Thread(target=start_parallel_containers, args=(self._external_containers, self.replay,))
+        required_thread.start()
+        external_thread = Thread(target=start_sequential_containers, args=(self._required_containers, self.replay))
```

Comment:
This code my have the weblog started before DB are up and healthy, some weblog interacts with DB at startup, so they must be healthy before we start weblog.

The more. I see this, the more I think we should introduce the notion of `need` in `TestedContainer`, create a compose.yaml, and let `docker compose` do the job. WDYT?

Reply:
> This code my have the weblog started before DB are up and healthy, some weblog interacts with DB at startup, so they must be healthy before we start weblog.

My main issue is that I'm not fully aware of what everything is doing and the actual interdependencies. I tried to separate things into 3 lists but that may have been incorrect or insufficient.

> The more. I see this, the more I think we should introduce the notion of need in TestedContainer, create a compose.yaml, and let docker compose do the job. WDYT?

I wish it was that simple, but unfortunately `docker compose` is unable to handle this. When using links or `depends_on`, it only makes sure to start the containers in order, but it doesn't have the notion of waiting for health checks, that is left to the applications. I usually agree with this design as applications should always be built with no specific expectation of ordering, but in this case that doesn't quite work as there is a lot of test setup involve that the applications don't know about.

With that said, I agree with the sentiment, and I was trying at first to come up with some sort of `ContainerGroup` that would allow defining containers that can be started in parallel and with a dependency system, but it ended up with too many code changes and I don't know Python well enough to make that big a change. I'll try to see if I can do something like this but while keeping changes minimum.

Reply:
Actually, would it be enough to just add `weblog` to the internal containers instead? I would expect weblog and the buddies to be able to be started in parallel since they may talk to each other and they all need to be healthy for requests to go through properly so there shouldn't be any ordering dependencies between them.

Reply:
> My main issue is that I'm not fully aware of what everything is doing and the actual interdependencies. I tried to separate things into 3 lists but that may have been incorrect or insufficient.

Let me give you some help on that, I'll write a small doc asap.

> I wish it was that simple, but unfortunately docker compose is unable to handle this.

Yes it can, you have a `condition: service_healthy` option in `depends_on` parameter. And on purpose, I kept all the healthcheck definition in the format of docker compose, because I knew one day we'll need to go into that direction.

>  I would expect weblog and the buddies to be able to be started in parallel since they may talk to each other and they all need to be healthy for requests to go through properly so there shouldn't be any ordering dependencies between them.

Agree.

Reply:
I didn't know about `service_healthy`, is it compatible with the health checks that are used in this project?

Reply:
> I didn't know about service_healthy

https://docs.docker.com/compose/startup-order/

> is it compatible with the health checks that are used in this project?

As said, I kept healthcheck definitions in the exact format of docker compose healthchecks : [Example](https://github.com/DataDog/system-tests/blob/main/utils/_context/containers.py#L634)

Reply:
> Let me give you some help on that, I'll write a small doc asap.

https://github.com/DataDog/system-tests/blob/main/docs/internals/end-to-end-life-cycle.md

Reply:
Any pointers on how I could switch this to Docker Compose?

Reply:
1. Add an optional `need:list[TestedContainer]` parameter to `TestedContainer`'s `__init__`. If not filled, it's initialized with an empty list.
2. add a method on `Scenario` class that build a  `compose.yml` base on `required_container`, with the good `need` value
3. on scenario warmup, spawn a blocking process, calling `docker compose` with the result of the function above as argument.
4. if it fails, print stdout/stderr

Reply:
And as the current code adds container in startup order, it'll require a small revamp to change this order, as weblog will mostly be in first.